experimental/overhead_matching/swag/evaluation/evaluate_swag.py

```diff
@@ -44,21 +47,132 @@ def evaluate_prediction_top_k(
+        # move
+        particle_state = sa.move_wag(particle_state, motion_delta, wag_config, generator)
+
+    # apply final observation
+    particle_state = sa.observe_wag(particle_state,
+                                    patch_similarity_for_path[-1],
+                                    satellite_patch_kdtree,
+                                    wag_config,
+                                    generator)
+    particle_history.append(particle_state.cpu().clone())
+    return particle_history
+
+
+def evaluate_model_on_paths(
+    vigor_dataset: vd.VigorDataset,
+    sat_model: torch.nn.Module,
+    pano_model: torch.nn.Module,
+    paths: list[list[int]],
+    wag_config: sa.WagConfig,
+    seed: int,
+    output_path: Path,
+    device: torch.device = "cuda:0",
+) -> None:
+
+    generator = torch.Generator(device=device).manual_seed(seed)
+    sat_data_view = vigor_dataset.get_sat_patch_view()
+    sat_data_view_loader = vd.get_dataloader(sat_data_view, batch_size=64, num_workers=16)
+    pano_data_view = vigor_dataset.get_pano_view()
+    pano_data_view_loader = vd.get_dataloader(pano_data_view, batch_size=64, num_workers=16)
+    print("building satellite embedding database")
+    with torch.no_grad():
+        sat_patch_positions = vigor_dataset.get_patch_positions().to(device)
+        sat_patch_kdtree = build_kd_tree(sat_patch_positions)
+        # sat_patch_kdtree = vigor_dataset._satellite_kdtree
+        if Path("/tmp/similarity_db.pt").exists():
+            all_similarity = torch.load("/tmp/similarity_db.pt").to(device)
+            print("loaded similarity db")
+        else:
+            sat_embeddings = sed.build_embeddings_from_model(
+                sat_model, sat_data_view_loader, lambda x: x.satellite, device=device)
+            print("building satellite embedding database done")
+            print("building panorama embedding database")
+            pano_embeddings = sed.build_embeddings_from_model(
+                pano_model, pano_data_view_loader, lambda x: x.panorama, device=device)
+            print("building all similarity")
+            all_similarity = sed.calculate_cos_similarity_against_database(
+                pano_embeddings, sat_embeddings)  # pano_embeddings x sat_patches
+
+            torch.save(all_similarity.cpu(), "/tmp/similarity_db.pt")
+
+        print("starting iter over paths")
+        for i, path in enumerate(tqdm.tqdm(paths)):
+            path_similarity_values = all_similarity[path]
+            motion_deltas = get_motion_deltas_from_path(vigor_dataset, path).to(device)
+            gt_initial_position_lat_lon = vigor_dataset._panorama_metadata.loc[path[0]]
+            gt_initial_position_lat_lon = torch.tensor(
+                (gt_initial_position_lat_lon['lat'], gt_initial_position_lat_lon['lon']), device=device)
+            initial_particle_state = sa.initialze_wag_particles(
+                gt_initial_position_lat_lon, wag_config, generator).to(device)
```

Comment:
Consider resetting the generator to the index of the path so that the order of evaluation doesn't matter. This means that you could split evaluation on multiple machines.